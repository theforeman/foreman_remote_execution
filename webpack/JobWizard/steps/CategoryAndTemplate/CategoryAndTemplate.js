import React from 'react';
import PropTypes from 'prop-types';
import { Title, Text, TextVariants, Form } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';
import { SelectField } from '../form/SelectField';
import { GroupedSelectField } from '../form/GroupedSelectField';

export const CategoryAndTemplate = ({
  jobCategories,
  jobTemplates,
  setJobTemplate,
  selectedTemplateID,
  selectedCategory,
  setCategory,
}) => {
  const templatesGroups = {};
  jobTemplates.forEach(template => {
    if (templatesGroups[template.provider_type]?.options)
      templatesGroups[template.provider_type].options.push({
        label: template.name,
        value: template.id,
      });
    else
      templatesGroups[template.provider_type] = {
        options: [{ label: template.name, value: template.id }],
        groupLabel: template.provider_type,
      };
  });

  const selectedTemplate = jobTemplates.find(
    template => template.id === selectedTemplateID
  )?.name;

  const onSelectCategory = newCategory => {
    setCategory(newCategory);
    setJobTemplate(null);
  };
  return (
    <>
      <Title headingLevel="h2">{__('Category And Template')}</Title>
      <Text component={TextVariants.p}>{__('All fields are required.')}</Text>
      <Form>
        <SelectField
          label={__('Job category')}
          fieldId="job_category"
          options={jobCategories}
          setValue={onSelectCategory}
          value={selectedCategory}
        />
        <GroupedSelectField
          label={__('Job template')}
          fieldId="job_template"
          groups={Object.values(templatesGroups)}
          setSelected={setJobTemplate}
          selected={selectedTemplate}
        />
      </Form>
    </>
  );
};

CategoryAndTemplate.propTypes = {
  jobCategories: PropTypes.array,
  jobTemplates: PropTypes.array,
  setJobTemplate: PropTypes.func.isRequired,
  selectedTemplateID: PropTypes.number,
  setCategory: PropTypes.func.isRequired,
  selectedCategory: PropTypes.string,
};
CategoryAndTemplate.defaultProps = {
  jobCategories: [],
  jobTemplates: [],
  selectedTemplateID: null,
  selectedCategory: null,
};

export default CategoryAndTemplate;
