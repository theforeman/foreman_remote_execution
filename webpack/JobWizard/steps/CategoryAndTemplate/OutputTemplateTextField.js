import React from 'react';
import { FormGroup, TextArea } from '@patternfly/react-core';
import PropTypes from 'prop-types';

export const OutputTemplateTextField = ({
  runtimeTemplate,
  setRuntimeTemplate,
}) => (
  <FormGroup label="Runtime template" isRequired={false}>
    <TextArea
      required={false}
      value={runtimeTemplate}
      onChange={setRuntimeTemplate}
    />
  </FormGroup>
);

OutputTemplateTextField.propTypes = {
  runtimeTemplate: PropTypes.string.isRequired,
  setRuntimeTemplate: PropTypes.func.isRequired,
};
