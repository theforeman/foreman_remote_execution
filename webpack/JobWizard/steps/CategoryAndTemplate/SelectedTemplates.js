import React from 'react';
import PropTypes from 'prop-types';
import { Chip, ChipGroup, Button } from '@patternfly/react-core';
import { translate as __ } from 'foremanReact/common/I18n';

const SelectedTemplate = ({ selected, setSelected, categoryName }) => {
  const deleteItem = itemToRemove => {
    setSelected(oldSelected =>
      oldSelected.filter(item => item.id !== itemToRemove)
    );
  };
  return (
    <ChipGroup className="templates-chip-group" categoryName={categoryName}>
      {selected.map(({ name, id }, index) => (
        <Chip
          key={index}
          id={`${categoryName}-${id}`}
          onClick={() => deleteItem(id)}
          closeBtnAriaLabel={`Close ${name}`}
        >
          {name}
        </Chip>
      ))}
    </ChipGroup>
  );
};

export const SelectedTemplates = ({
  selectedOutputTemplates,
  setOutputTemplates,
  runtimeTemplate,
  setRuntimeTemplate,
}) => {
  const clearAll = () => {
    setOutputTemplates(() => []);
    setRuntimeTemplate(() => '');
  };
  return (
    <div className="selected-chips">
      <SelectedTemplate
        selected={selectedOutputTemplates}
        categoryName="Predefined templates"
        setSelected={setOutputTemplates}
      />
      <Button variant="link" className="clear-chips" onClick={clearAll}>
        {__('Clear templates')}
      </Button>
    </div>
  );
};

SelectedTemplates.propTypes = {
  selectedOutputTemplates: PropTypes.array.isRequired,
  setOutputTemplates: PropTypes.func.isRequired,
  runtimeTemplate: PropTypes.string.isRequired,
  setRuntimeTemplate: PropTypes.func.isRequired,
};

SelectedTemplate.propTypes = {
  categoryName: PropTypes.string.isRequired,
  selected: PropTypes.array.isRequired,
  setSelected: PropTypes.func.isRequired,
};
