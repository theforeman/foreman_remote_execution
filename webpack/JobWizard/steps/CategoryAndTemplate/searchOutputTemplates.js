import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import URI from 'urijs';
import { SelectVariant } from '@patternfly/react-core';
import { get } from 'foremanReact/redux/API';
import { selectResponse, selectIsLoading } from '../../JobWizardSelectors';
import { SearchSelect } from '../form/SearchSelect';

export const searchOutputTemplates = (apiKey, url) => {
  const dispatch = useDispatch();
  const uri = new URI(url);
  const onSearch = search =>
    dispatch(
      get({
        key: apiKey,
        url: uri.addSearch({
          per_page: 'all',
        }),
      })
    );

  const response = useSelector(state => selectResponse(state, apiKey));
  const isLoading = useSelector(state => selectIsLoading(state, apiKey));
  return [onSearch, response, isLoading];
};

export const OutputSelect = props => (
  <SearchSelect
    {...props}
    variant={SelectVariant.typeaheadMulti}
    useNameSearch={searchOutputTemplates}
  />
);
