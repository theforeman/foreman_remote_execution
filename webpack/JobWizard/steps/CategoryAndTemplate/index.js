import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import URI from 'urijs';
import { get } from 'foremanReact/redux/API';
import {
  selectJobCategories,
  selectJobTemplates,
  selectJobCategoriesStatus,
  filterJobTemplates,
} from '../../JobWizardSelectors';
import { CategoryAndTemplate } from './CategoryAndTemplate';

import {
  JOB_TEMPLATES,
  JOB_CATEGORIES,
  templatesUrl,
} from '../../JobWizardConstants';

const ConnectedCategoryAndTemplate = ({
  jobTemplate,
  setJobTemplate,
  category,
  setCategory,
}) => {
  const dispatch = useDispatch();

  const jobCategoriesStatus = useSelector(selectJobCategoriesStatus);
  useEffect(() => {
    if (!jobCategoriesStatus) {
      // Don't reload categories if they are already loaded
      dispatch(
        get({
          key: JOB_CATEGORIES,
          url: '/ui_job_wizard/categories',
          handleSuccess: response =>
            setCategory(response.data.job_categories[0] || ''),
        })
      );
    }
  }, [jobCategoriesStatus, dispatch, setCategory]);

  const jobCategories = useSelector(selectJobCategories);

  useEffect(() => {
    if (category) {
      const templatesUrlObject = new URI(templatesUrl);
      dispatch(
        get({
          key: JOB_TEMPLATES,
          url: templatesUrlObject.addSearch({
            search: `job_category="${category}"`,
            per_page: 'all',
          }),
          handleSuccess: response =>
            setJobTemplate(
              Number(filterJobTemplates(response?.data?.results)[0]?.id) || null
            ),
        })
      );
    }
  }, [category, dispatch, setJobTemplate]);

  const jobTemplates = useSelector(selectJobTemplates);

  return (
    <CategoryAndTemplate
      jobTemplates={jobTemplates}
      jobCategories={jobCategories}
      setJobTemplate={setJobTemplate}
      selectedTemplateID={jobTemplate}
      setCategory={setCategory}
      selectedCategory={category}
    />
  );
};

ConnectedCategoryAndTemplate.propTypes = {
  jobTemplate: PropTypes.number,
  setJobTemplate: PropTypes.func.isRequired,
  category: PropTypes.string.isRequired,
  setCategory: PropTypes.func.isRequired,
};
ConnectedCategoryAndTemplate.defaultProps = { jobTemplate: null };

export default ConnectedCategoryAndTemplate;
