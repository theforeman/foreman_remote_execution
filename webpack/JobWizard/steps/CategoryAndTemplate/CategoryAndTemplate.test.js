import { testComponentSnapshotsWithFixtures } from '@theforeman/test';
import { CategoryAndTemplate } from './CategoryAndTemplate';

const fixtures = {
  'renders with props': {
    jobCategories: [
      'Commands',
      'Ansible Playbook',
      'Ansible Galaxy',
      'Ansible Roles Installation',
    ],
    jobTemplates: [
      {
        id: 190,
        name: 'ab Run Command - SSH Default clone',
        job_category: 'Commands',
        provider_type: 'SSH',
        snippet: false,
      },
      {
        id: 168,
        name: 'Ansible Roles - Ansible Default',
        job_category: 'Ansible Playbook',
        provider_type: 'Ansible',
        snippet: false,
      },
      {
        id: 170,
        name: 'Ansible Roles - Install from git',
        job_category: 'Ansible Roles Installation',
        provider_type: 'Ansible',
        snippet: false,
      },
    ],
    setJobTemplate: jest.fn(),
    selectedTemplateID: 190,
    setCategory: jest.fn(),
    selectedCategory: 'I am a category',
  },
};

describe('CategoryAndTemplate', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(CategoryAndTemplate, fixtures));
});
