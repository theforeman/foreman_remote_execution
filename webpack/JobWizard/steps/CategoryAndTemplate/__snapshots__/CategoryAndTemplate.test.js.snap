// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryAndTemplate rendering renders with props 1`] = `
<Fragment>
  <Title
    headingLevel="h2"
  >
    Category And Template
  </Title>
  <Text
    component="p"
  >
    All fields are required.
  </Text>
  <Form>
    <SelectField
      fieldId="job_category"
      label="Job category"
      options={
        Array [
          "Commands",
          "Ansible Playbook",
          "Ansible Galaxy",
          "Ansible Roles Installation",
        ]
      }
      setValue={[Function]}
      value="I am a category"
    />
    <GroupedSelectField
      fieldId="job_template"
      groups={
        Array [
          Object {
            "groupLabel": "SSH",
            "options": Array [
              Object {
                "label": "ab Run Command - SSH Default clone",
                "value": 190,
              },
            ],
          },
          Object {
            "groupLabel": "Ansible",
            "options": Array [
              Object {
                "label": "Ansible Roles - Ansible Default",
                "value": 168,
              },
              Object {
                "label": "Ansible Roles - Install from git",
                "value": 170,
              },
            ],
          },
        ]
      }
      label="Job template"
      selected="ab Run Command - SSH Default clone"
      setSelected={[MockFunction]}
    />
  </Form>
</Fragment>
`;
